src/pybind/mgr/dashboard/controllers/auth.py

```diff
@@ -28,33 +28,41 @@ class Auth(RESTController):
     """
     Provide authenticates and returns JWT token.
     """
-
     def create(self, username, password):
         user_data = AuthManager.authenticate(username, password)
         user_perms, pwd_expiration_date, pwd_update_required = None, None, None
-        if user_data:
-            user_perms = user_data.get('permissions')
-            pwd_expiration_date = user_data.get('pwdExpirationDate', None)
-            pwd_update_required = user_data.get('pwdUpdateRequired', False)
-
-        if user_perms is not None:
-            logger.debug('Login successful')
-            token = JwtManager.gen_token(username)
-            token = token.decode('utf-8')
-            cherrypy.response.headers['Authorization'] = "Bearer: {}".format(token)
-            return {
-                'token': token,
-                'username': username,
-                'permissions': user_perms,
-                'pwdExpirationDate': pwd_expiration_date,
-                'sso': mgr.SSO_DB.protocol == 'saml2',
-                'pwdUpdateRequired': pwd_update_required
-            }
+        attempt = mgr.ACCESS_CTRL_DB.get_attempt(username)
+        if attempt < 10:
+            if user_data:
+                user_perms = user_data.get('permissions')
+                pwd_expiration_date = user_data.get('pwdExpirationDate', None)
+                pwd_update_required = user_data.get('pwdUpdateRequired', False)
 
-        logger.debug('Login failed')
-        raise DashboardException(msg='Invalid credentials',
-                                 code='invalid_credentials',
-                                 component='auth')
+            if user_perms is not None:
+                logger.debug('Login successful')
+                attempt = 0
+                mgr.ACCESS_CTRL_DB.save_attempt(username, attempt)
+                token = JwtManager.gen_token(username)
+                token = token.decode('utf-8')
+                cherrypy.response.headers['Authorization'] = "Bearer: {}".format(token)
+                return {
+                    'token': token,
+                    'username': username,
+                    'permissions': user_perms,
+                    'pwdExpirationDate': pwd_expiration_date,
+                    'sso': mgr.SSO_DB.protocol == 'saml2',
+                    'pwdUpdateRequired': pwd_update_required
+                }        
+            attempt += 1
+            mgr.ACCESS_CTRL_DB.save_attempt(username, attempt)
+            logger.debug('Login failed %s %s', username, attempt)
+            raise DashboardException(msg='Invalid credentials',
+                                    code='invalid_credentials',
+                                    component='auth')  
+        else:
+            user = mgr.ACCESS_CTRL_DB.get_user(username)
```

Comment:
IMO it would make sense to log the failed log in attempt (user is banned) so that the administrator can see that in the logs.

Reply:
Done.